zhaquirks/danfoss/thermostat.py

```diff
@@ -26,113 +55,442 @@
+            id=0x404F, type=types.Bool, access="rp", is_manufacturer_specific=True
+        )
+        preheat_time = ZCLAttributeDef(
+            id=0x4050, type=types.uint32_t, access="rp", is_manufacturer_specific=True
+        )
+        window_open_feature = ZCLAttributeDef(  # window_open_feature_on_off
+            id=0x4051, type=types.Bool, access="rw", is_manufacturer_specific=True
+        )  # non-configurable reporting
 
     async def write_attributes(self, attributes, manufacturer=None):
-        """Send SETPOINT_COMMAND after setpoint change."""
+        """There are 2 types of setpoint changes: Fast and Slow.
+        Fast is used for immediate changes; this is done using a command (setpoint_command).
+        Slow is used for scheduled changes; this is done using an attribute (occupied_heating_setpoint).
+        In case of a change on occupied_heating_setpoint, a setpoint_command is used.
+
+        Thermostatic radiator valves from Danfoss cannot be turned off to prevent damage during frost.
+        This is emulated by setting setpoint to the minimum setpoint.
+        """
+
+        fast_setpoint_change = None
+
+        if occupied_heating_setpoint.name in attributes:
+            # Store setpoint for use in command
+            fast_setpoint_change = attributes[occupied_heating_setpoint.name]
 
+        if attributes.get(system_mode.name) == system_mode.type.Off:
+            # Just turn setpoint down to minimum temperature using fast_setpoint_change
+            fast_setpoint_change = self._attr_cache[min_heat_setpoint_limit.id]
+            attributes[occupied_heating_setpoint.name] = fast_setpoint_change
+            attributes[system_mode.name] = system_mode.type.Heat
+
+        # Attributes cannot be empty, because write_res cannot be empty, but it can contain unrequested items
         write_res = await super().write_attributes(
             attributes, manufacturer=manufacturer
         )
 
-        if "occupied_heating_setpoint" in attributes:
-            self.debug(
-                "sending setpoint command: %s", attributes["occupied_heating_setpoint"]
-            )
+        if fast_setpoint_change is not None:
+            # On Danfoss a fast setpoint change is done through a command
             await self.setpoint_command(
-                0x01, attributes["occupied_heating_setpoint"], manufacturer=manufacturer
+                DanfossSetpointCommandEnum.User_interaction,
+                fast_setpoint_change,
+                manufacturer=manufacturer,
             )
 
         return write_res
 
+    async def bind(self):
+        """According to the documentation of Zigbee2MQTT there is a bug in the Danfoss firmware with the time.
+        It doesn't request it, so it has to be fed the correct time.
+        """
+        await self.endpoint.time.write_time()
 
-class DanfossUserInterfaceCluster(CustomCluster, UserInterface):
-    """Danfoss custom cluster."""
+        return await super().bind()
```

Comment:
Since we're calling the `bind()` method for the Time cluster in the tests and the `bind()` method for the Thermostat cluster isn't covered yet, you can just call that method in the `test_danfoss_time_bind` test.

Reply:
You're right.